packages/gatsby/src/redux/run-sift.js

```diff
@@ -354,22 +450,37 @@ const filterToStats = (
  *
  * @param {Array<DbQuery>} filters Resolved. (Should be checked by caller to exist)
  * @param {Array<string>} nodeTypeNames
- * @param {Map<FilterCacheKey, Map<string | number | boolean, Set<IGatsbyNode>>>} typedKeyValueIndexes
- * @returns {Array|undefined} Collection of results
+ * @param {undefined | null | FiltersCache} filtersCache
+ * @returns {Array<IGatsbyNode> | undefined} Collection of results
  */
-const filterWithoutSift = (filters, nodeTypeNames, typedKeyValueIndexes) => {
+const filterWithoutSift = (filters, nodeTypeNames, filtersCache) => {
   // This can also be `$ne`, `$in` or any other grapqhl comparison op
+
+  if (!filtersCache) {
+    // If no filter cache is passed on, explicitly don't use one
+    return undefined
+  }
+
+  if (filters.length === 0) {
+    // If no filters are given, go through Sift. This does not appear to be
+    // slower than s
+    // hortcutting it here.
+    return undefined
+  }
+
   if (
-    !typedKeyValueIndexes ||
-    filters.length === 0 || // TODO: we should special case this
     filters.some(
-      filter => filter.type === `elemMatch` || filter.query.comparator !== `$eq`
+      filter =>
+        filter.type === `query` && // enabled
+        // filter.type === `elemMatch` || // disabled
```

Comment:
This was the simple toggle to enable/disable support for `elemMatch`.

Yes, if the fast indexes fail in any way (`undefined` or empty set) it has to go through Sift. This is because there are some edge cases that the fast indexes can't support yet (like `@proxy`) and we don't want to accidentally regress.

I tried to have the fast indexes fail as fast as possible.